scripts/qa_crowbarsetup.sh

```diff
@@ -5038,10 +5039,140 @@ function ping_fips
+        done
+
+    rm $all_fips
+}
+
+
+function ping_stack_fips
+{
+    . .openrc
+
+    local stack=$1
+    local fips=$(mktemp)
+
+    get_stack_fips $stack $fips
+
+    for fip in $(cat $fips)
+        do
+        ping -c 1 -w 60 $fip || complain 120 "cannot reach stack $stack floating IP $fip."
+        done
+
+    rm $fips
+}
+
+# Disable all but 3 compute nodes (these are the ones we will deploy the
+# pre-upgrade Heat stack on)
+function oncontroller_reduce_to_3_computes
+{
+    . .openrc
+    echo "Reducing cloud to 3 active compute nodes..."
+    local enable_computes=$(nova service-list | grep nova-compute | awk '{print $2}' | head -n 3)
+    local disable_computes=$(nova service-list | grep nova-compute | awk '{print $2}' | tail -n +4)
+
+    echo "Nodes to enable (first 3):"
+    for node in $enable_computes; do
+        echo "  $node"
+    done
+
+    echo "Nodes to disable: $disable_computes"
+    for node in $disable_computes; do
+        echo "  $node"
+    done
+
+    echo
+    echo "Enabling first 3 nodes..."
+
+    for service in $enable_computes; do
+        nova service-enable $service
+    done
+
+    echo "First 3 nodes enabled."
+
+    echo "Disabling remaining nodes..."
+
+    for service in $disable_computes; do
+        nova service-disable $service
+    done
+
+    echo "Remaining nodes disabled."
+}
+
+# Re-enable the compute nodes we disabled
+function oncontroller_reenable_computes
+{
+    . .openrc
+
+    echo "Re-enabling disabled compute nodes..."
+    for service in $(nova service-list | grep nova-compute | grep disabled | awk '{print $2}'); do
+        nova service-enable $service
+    done
+}
+
+# Use heat_stack_params to provide parameters to heat template
+function oncontroller_testpreupgrade_extended
+{
+    echo "Running extended pre-upgrade test..."
+
+    local tempest_image=$(openstack image list -f value | cut -d ' ' -f 2 | grep '^cirros-.*-tempest$' | head -n 1)
+    local nodes_enabled=$(nova service-list | grep nova-compute | grep enabled | awk '{print $2}')
+    local enabled_count=$(echo $nodes_enabled | wc -w)
+    if [ $enabled_count -lt 3 ]; then
+        complain 11 "testpreupgrade_extended: not enough compute nodes (${enabled_count}). please ensure there are at least 3 enabled compute nodes."
+    fi
+
+    if [ -z "$tempest_image" ]; then
+        # Depending on URL in the tempest barclamp, the image's name may differ
+        # from the one matched by the regex above. In that case the regex below
+        # should match:
+        tempest_image=$(openstack image list -f value | cut -d ' ' -f 2 | grep '^cirros-.*-tempest-machine$' | head -n 1)
+        if [ -z "$tempest_image" ]; then
+            complain 11 "No tempest image found. Please make sure the tempest barclamp created its images."
+        fi
+    fi
+
+    for i in $(seq 1 3); do
+        local template=$(echo $extended_upgrade_templates | cut -d ' ' -f $i)
```

Comment:
Wow, so that's how you iterate over lists in bash? ;-)

Reply:
One could use bash arrays for that but I'm never too sure to what extent they are a bash-ism or `sh`compatible so I stick with this - admittedly clumsy - approach :-)

Reply:
Yeah, bash arrays are even uglier to read...